content/en/blog/_posts/2024-03-19-go-workspaces.md

```diff
@@ -0,0 +1,186 @@
+compiling and running tests with workspaces enabled. I set out to purge the
+repository of anything GOPATH related. That's when I hit the first real bump -
+the code-generation tools.
+
+We had about a dozen tools, totalling several thousand lines of code. All of
+them were built using an internal framework called
+[gengo](https://github.com/kubernetes/gengo), which was built on Go's own
+parsing libraries. There were 2 main problems:
+
+  1) Those parsing libraries didn't understand modules or workspaces.
+  2) GOPATH allowed us to pretend that Go "package paths" and directories on
+     disk were interchangeable in trivial ways. They are not.
+
+Switching to a [modules- and workspaces-aware parsing](https://pkg.go.dev/golang.org/x/tools@v0.19.0/go/packages)
+library was the first step. Then I had to make a long series of changes to
+each of the code-generation tools. Critically, I had to find a way to do it
+that was possible for some other person to review! After some conversations
+with likely reviewers, we agreed that a series of small commits would be
+easiest to review.
+
+Sadly, these were not mechanical changes. I had to dig into each tool to
+figure out where they were processing disk paths versus where they were
+processing package names, and where those were being conflated. I made
+extensive use of the [delve](https://github.com/go-delve/delve) debugger, which
+I just can't say enough good things about.
+
+One unfortunate result of this work was that I had to break compatibility. The
+gengo library simply did not have enough information to process packages
+outside of GOPATH. After discussion with gengo and Kubernetes maintainers, we
+agreed to make [gengo/v2](https://github.com/kubernetes/gengo/tree/master/v2).
+I also used this as an opportunity to clean up some of the gengo APIs and the
+tools' CLIs to be more understandable and not conflate packages and
+directories, for example you can't just string-join directory names and
+assume the result is a valid package name.
+
+Once I had the code-generation tools converted, I shifted attention to the
+dozens of scripts in the `hack` directory. One by one I had to run them, debug,
+and fix failures. Some of them needed minor changes and some needed to be
+rewritten.
+
+Along the way we hit some cases that Go did not support, like workspace
+vendoring. Kubernetes depends on vendoring to ensure that our dependencies are
+always available, even if their source code is removed from the internet (it
+has happened more than once!). After discussing with the Go team, and looking
+at possible workarounds, they decided the right path was to
+[implement workspace vendoring](https://github.com/golang/go/issues/60056).
+
+The eventual Pull Request contained over 200 individual commits.
+
+## Results
+
+Now that this work has been merged, what does this mean for Kubernetes users?
+Pretty much nothing. No features were added or changed. This work was not
```

Comment:
So while I get that workspaces helps us as maintainers, this line might make readers question why we're making such a big deal about this with a blog. What do you think about adding a call to action about getting involved in the project and working up the contribution ladder, if folks are interested in experiencing our now simpler and more consistent codebase?

Reply:
Heard.  I wanted to emphasize that this work - hidden, dirty work - is necessary and valued, and that "even long time maintainers" get dirty.  THAT is the point of the blog, really.  I'll think about how to make a call-to-action.

Reply:
I added a few sentences about "technical debt" - feedback welcome!

Reply:
I think what you've added is great, thank you! The focus on growth and evolution of the project as an indication of why this was needed is spot on.